src/playlist.rs

```diff
@@ -1144,6 +1225,253 @@ impl DateRange {
+            .parse::<f64>()
+            .map_err(|err| format!("Failed to parse PART-HOLD-BACK attribute: {}", err)));
+        let can_block_reload = is_yes!(attrs, "CAN-BLOCK-RELOAD");
+
+        Ok(ServerControl {
+            can_skip_until,
+            can_skip_dateranges,
+            hold_back,
+            part_hold_back,
+            can_block_reload,
+        })
+    }
+
+    pub(crate) fn write_to<T: Write>(&self, w: &mut T) -> std::io::Result<()> {
+        write!(w, "#EXT-X-SERVER-CONTROL:")?;
+        write_some_float_attribute!(w, "CAN-SKIP-UNTIL", &self.can_skip_until)?;
+        if self.can_skip_dateranges {
+            write!(w, ",CAN-SKIP-DATERANGES=YES")?;
+        }
+        write_some_float_attribute!(w, ",HOLD-BACK", &self.hold_back)?;
+        write_some_float_attribute!(w, ",PART-HOLD-BACK", &self.part_hold_back)?;
+        if self.can_block_reload {
+            write!(w, ",CAN-BLOCK-RELOAD=YES")?;
+        }
+        writeln!(w)
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Clone)]
+pub struct PartInf {
+    pub part_target: f64,
+}
+
+impl PartInf {
+    pub(crate) fn from_hashmap(
+        mut attrs: HashMap<String, QuotedOrUnquoted>,
+    ) -> Result<PartInf, String> {
+        let part_target = unquoted_string_parse!(attrs, "PART-TARGET", |s: &str| s
+            .parse::<f64>()
+            .map_err(|err| format!("Failed to parse PART-TARGET attribute: {}", err)))
+        .ok_or_else(|| String::from("EXT-X-PART-INF without mandatory PART-TARGET attribute"))?;
+
+        Ok(PartInf { part_target })
+    }
+
+    pub(crate) fn write_to<T: Write>(&self, w: &mut T) -> std::io::Result<()> {
+        match WRITE_OPT_FLOAT_PRECISION.load(Ordering::Relaxed) {
+            MAX => {
+                write!(w, "#EXT-X-PART-INF:PART-TARGET={}", self.part_target)?;
+            }
+            n => {
+                write!(w, "#EXT-X-PART-INF:PART-TARGET={:.*}", n, self.part_target)?;
+            }
+        };
+
+        writeln!(w)
+    }
+}
+
+#[derive(Debug, Default, PartialEq, Clone)]
+pub struct Part {
+    pub uri: String,
+    pub duration: f64,
+    pub independent: bool,
+    pub gap: bool,
+    pub byte_range: Option<ByteRange>,
+}
+
+impl Part {
+    pub(crate) fn from_hashmap(
+        mut attrs: HashMap<String, QuotedOrUnquoted>,
+    ) -> Result<Part, String> {
+        let uri = quoted_string!(attrs, "URI")
+            .ok_or_else(|| String::from("EXT-X-PART without mandatory URI attribute"))?;
+        let duration = unquoted_string_parse!(attrs, "DURATION", |s: &str| s
+            .parse::<f64>()
+            .map_err(|err| format!("Failed to parse DURATION attribute: {}", err)))
+        .ok_or_else(|| String::from("EXT-X-PART without mandatory DURATION attribute"))?;
+        let independent = is_yes!(attrs, "INDEPENDENT");
+        let gap = is_yes!(attrs, "GAP");
+        let byte_range = quoted_string_parse!(attrs, "BYTERANGE", |s: &str| s.parse::<ByteRange>());
```

Comment:
The function from the parser might still be reused like this:

```
let byte_range = quoted_string_parse!(attrs, "BYTERANGE", |s: &str| {
            byte_range_val(s.as_bytes())
                .finish()
                .map(|(_, byterange)| byterange)
                .map_err(|_| format!("Failed to parse BYTERANGE attribute"))
        });
```

Then the `impl Display` and `impl FromStr` wouldn't be needed. 

Fine either way though.

Reply:
Well I considered going that way around we would be depending on the parser from the playlist, which is normally guarded behind the parser feature. I figured not having those coupled was an intentional design decision?

Reply:
Good point, it should probably not depend on nom. I see there is already a reference to the `QuotedOrUnquoted` struct in the parser, not sure how that works actually.